### ISSUE TITLE
Integrating pre-trained embeddings & Using input_fn inside collect_concepts

### ISSUE THREAD

**Issue by benediktwagner** [fursovia/tcav_nlp#1]
Hi Ivan, first of all thank you for sharing your code.
I am starting to run a few experiments and wanted to integrate pre-trained embeddings. 

So far I was thinking about adding it into the model_fn and use the nn.embedding_lookup.
Alternatively one could also integrate it directly inside the vectorize function inside the input_fn.py

The following is the snippet I am using to load in embeddings. 
```
def loadPretrainedModel(file):
    print("Loading Pretrained Model")
    f = open(file,'r')
    model = {}
    for line in f:
        splitLine = line.split()
        word = splitLine[0]
        embedding = np.array([float(val) for val in splitLine[1:]])
        model[word] = embedding
    print("Done.",len(model)," words loaded!")
    return model
```

I was considering something like the following: 
```
word_model = loadPretrainedModel(file):
variable = tf.Variable(word_model, dtype=tf.float32, trainable=False)
embeddings = tf.nn.embedding_lookup(variable, features['x'], name='emb_matrix_lookup')
```
Somehow this is not really working though. 

Alternatively one could integrate the gensim library which could speed up things perhaps.
I would really appreciate any thoughts you have on this. 

––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––

In addition to this, I wanted to ask whether you think it is possible to use the input_fn.py mapping for getting the gradients inside the collect_concepts.py too. 

Currently the code is using
```
grads = mw.calculate_grad(sess, labels, train['text'].tolist())
```
However, the issue is that this doesn't work with input sequences longer than 10 as it doesn't automatically create the sequences. 
One way to get around is to transform the whole data into sequences that follow seq length but ideally one could use the same mapping from input_fn used for training the model. If you have any ideas how this could work with TensorFlow’s dataloader/mapping that would be great.  

In any case: thanks again for sharing and I would apreciate any help you could provide.

**Comment by fursovia** [comment]
Hi Ben! @benediktwagner 

One of the ways to initialize embedding matrix is to pass `initializer` to `tf.get_variable`. The following code should work:

```python
vocab = {0: 'a', 1: 'apple', ..., 9999: 'zebra'}
vocab_size = len(vocab)
embedding_size = 300

# you should create your own embedding matrix (fasttext/word2vec/glove/etc)
# each row of this matrix should correspond to an embedding vector of the corresponding word
# e.g. embedding_matrix[9999] is an embedding vector of `zebra` 
embedding_matrix = np.random.normal(0, 0.001, (vocab_size, embedding_size))

initializer = tf.constant_initializer(embedding_matrix)
# next replace this line
# https://github.com/fursovia/tcav_nlp/blob/master/model/model_fn.py#L44
emb_matrix = tf.get_variable(..., initializer=initializer)
```
Do not forget to handle `<PAD>` and `<UNK>` tokens. As I remember, the index of `<UNK>` is `vocab_size + 1`. You can initialize embeddings for these tokens randomly (e.g. truncated normal distribution). Everything else should remain the same.

____
Could you please explain why `...doesn't work with input sequences longer than 10...`? Did you know that you can change this hyper-parameter? (https://github.com/fursovia/tcav_nlp/blob/master/experiments/config.yaml#L3) Setting the value to the maximum length of sequences should work, however, this is not an efficient solution. Another solution is to change `padded_shapes` (https://github.com/fursovia/tcav_nlp/blob/master/model/input_fn.py#L43) to 

```python
padded_shapes = (tf.TensorShape([None]), tf.TensorShape([]))
```

Setting `padded_shapes` to `None` should pad all samples in the batch to the size of the largest element in the batch.

**Comment by benediktwagner** [comment]
Hi Ivan @fursovia! Thanks so much for your help!
Just to document: `<PAD>` is index `0` and `<UNK>` index `1` inside the `vocab.txt`. 
I did exactly what you said and it seems to be working. Although it did not change anything accuracy-wise in my data. I will try to check whether it actually worked and is using the correct embeddings. Just to make sure by replacing the following line, you meant to say that I would need to add the initializer argument to the statement (see below)?
```    
# replace this line
# https://github.com/fursovia/tcav_nlp/blob/master/model/model_fn.py#L44   
emb_matrix = tf.get_variable('embedding_matrix', shape=[params['vocab_size'], 
                             params['emb_dim']],dtype=tf.float64, initializer=initializer)
```


***
About the second question: I know that sequence lengths can be changed and I am not taking about the model itself. The question is only about the following computation of gradients.  
<https://github.com/fursovia/tcav_nlp/blob/master/collect_concepts.py#L117>
The issue here is that we vectorise separately on the data without the pre-processing from the input_fn.py. This means that for computing the gradients, we are not using padded shapes anymore but still need to fit the input `train['text'].tolist()` onto vectors with the pre-defined sequence length. So the idea would be to introduce padded_shapes somewhere inside the `get_input()` perhaps? <https://github.com/fursovia/tcav_nlp/blob/master/model/model_fn.py#L136> 

I hope this should be clearer now. Thanks again for your support – I really appreciate it.

**Comment by fursovia** [comment]
@benediktwagner 

`...by replacing the following line, you meant to say that I would need to add the initializer argument to the statement (see below)?` -- exactly. You did it right. 

____
Actually, you can use `input_fn` function just as you would use it during the training process. The following steps should work:

1. Save data to `.csv` format (like here https://github.com/fursovia/tcav_nlp/blob/master/data_prep.py#L51) (but do not overwrite `vocab.txt` file)
2. Then use `input_fn` to get a `tf.data.Dataset` object. (https://github.com/fursovia/tcav_nlp/blob/master/model/input_fn.py#L28)
3. Initialize an iterator from this object (https://www.tensorflow.org/versions/r1.12/api_docs/python/tf/data/Dataset#make_initializable_iterator)
4. Then use `get_next` method to iterate over the dataset (https://www.tensorflow.org/versions/r1.12/api_docs/python/tf/data/Iterator#get_next)

[This blog post](https://medium.com/ymedialabs-innovation/how-to-use-dataset-and-iterators-in-tensorflow-with-code-samples-3bb98b6b74ab) should explain how tensorflow iterators and datasets work.
